File: src/cpp/lib/test/CacheTestQObject.cpp
#include "test/CacheTestQObject.h"

CacheTestQObject::CacheTestQObject() : _foo(0), _bar(0) {}

CacheTestQObject* CacheTestQObject::foo() {
  if (_foo) {
    return _foo;
  }
  _foo = new CacheTestQObject();
  _foo->setParent(this);
  return _foo;
}

void CacheTestQObject::clearFoo() {
  if (_foo) {
    delete _foo;
    _foo = nullptr;
  }
}

CacheTestQObject* CacheTestQObject::bar() {
  if (_bar) {
    return _bar;
  }
  _bar = new CacheTestQObject();
  _bar->setParent(this);
  return _bar;
}

File: src/lib/core/__test__/CacheTestQObject.ts
import addon from '../../utils/addon';
import { NativeElement } from '../Component';
import { QObject, QObjectSignals } from '../../QtCore/QObject';
import { wrapperCache } from '../../core/WrapperCache';

export class CacheTestQObject extends QObject<QObjectSignals> {
    constructor(arg?: NativeElement) {
        let native;
        if (arg == null) {
            native = new addon.CacheTestQObject();
        } else {
            native = arg;
        }
        super(native);
    }

    foo(): CacheTestQObject {
        return wrapperCache.getWrapper(this.native.foo()) as CacheTestQObject;
    }

    clearFoo(): void {
        this.native.clearFoo();
    }

    bar(): CacheTestQObject {
        return wrapperCache.getWrapper(this.native.bar()) as CacheTestQObject;
    }
}
wrapperCache.registerWrapper('CacheTestQObjectWrap', CacheTestQObject);
